registry/storage/shares.go

```diff
@@ -210,48 +210,57 @@ func (s *sharesStorage) Save(rw basedb.ReadWriter, shares ...*types.SSVShare) er
 		}
 	}
 
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.logger.Debug("save validators to shares storage", zap.Int("count", len(shares)))
 
-	return s.unsafeSave(rw, shares...)
-}
+	s.storageMtx.Lock()
+	defer s.storageMtx.Unlock()
 
-func (s *sharesStorage) unsafeSave(rw basedb.ReadWriter, shares ...*types.SSVShare) error {
-	err := s.db.Using(rw).SetMany(s.prefix, len(shares), func(i int) (basedb.Obj, error) {
-		share := specShareToStorageShare(shares[i])
-		value, err := share.Encode()
-		if err != nil {
-			return basedb.Obj{}, fmt.Errorf("failed to serialize share: %w", err)
-		}
-		return basedb.Obj{Key: s.storageKey(share.ValidatorPubKey[:]), Value: value}, nil
-	})
-	if err != nil {
-		return err
-	}
+	// Update in-memory.
+	err := func() error {
+		s.memoryMtx.Lock()
+		defer s.memoryMtx.Unlock()
 
-	updateShares := make([]*types.SSVShare, 0, len(shares))
-	addShares := make([]*types.SSVShare, 0, len(shares))
+		updateShares := make([]*types.SSVShare, 0, len(shares))
+		addShares := make([]*types.SSVShare, 0, len(shares))
 
-	for _, share := range shares {
-		key := hex.EncodeToString(share.ValidatorPubKey[:])
+		for _, share := range shares {
+			key := hex.EncodeToString(share.ValidatorPubKey[:])
 
-		// Update validatorStore indices.
-		if _, ok := s.shares[key]; ok {
-			updateShares = append(updateShares, share)
-		} else {
-			addShares = append(addShares, share)
+			// Update validatorStore indices.
+			if _, ok := s.shares[key]; ok {
+				updateShares = append(updateShares, share)
+			} else {
+				addShares = append(addShares, share)
+			}
+			s.shares[key] = share
 		}
-		s.shares[key] = share
-	}
 
-	if err := s.validatorStore.handleSharesUpdated(updateShares...); err != nil {
-		return err
-	}
-	if err := s.validatorStore.handleSharesAdded(addShares...); err != nil {
+		if err := s.validatorStore.handleSharesUpdated(updateShares...); err != nil {
+			return err
+		}
+
+		if err := s.validatorStore.handleSharesAdded(addShares...); err != nil {
+			return err
+		}
```

Comment:
Would be nice to wrap errors here (because they come from different code branches might be hard to debug without it):
```go
		if err := s.validatorStore.handleSharesUpdated(updateShares...); err != nil {
			return fmt.Erorrf("handleSharesUpdated: %w", err)
		}

		if err := s.validatorStore.handleSharesAdded(addShares...); err != nil {
			return fmt.Erorrf("handleSharesAdded: %w", err)
		}
```